File: 3-longest-substring-without-repeating-characters/longest-substring-without-repeating-characters.py
class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:
        if not s:
            return 0
        mp = {}
        res = 0
        L = 0
        for R in range(len(s)):
            c = s[R]

            if c in mp:
                L = max(mp[c], L)
            
            res = max(res, R - L + 1)
            mp[c] = R + 1
        return res

File: 3-longest-substring-without-repeating-characters/longest-substring-without-repeating-characters.cpp
class Solution {
public:
    int lengthOfLongestSubstring(string s) {
        if (s.empty()) {
            return 0;
        }
        int res = 0, L = 0;
        unordered_map<char, int> counts;

        for(int R = 0; R < s.size(); ++R) {
            char c = s[R];
            counts[c]++;

            while (counts[c] > 1) {
                counts[s[L]]--;
                if (!counts[s[L]]) {
                    counts.erase(s[L]);
                }
                L++;
            }

            res = max(res, R - L + 1);
        }

        return res;

    }
};